File: near/src/main/java/com/adroitandroid/near/connect/server/TcpServerService.java
package com.adroitandroid.near.connect.server;

import android.annotation.SuppressLint;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Binder;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.IBinder;
import android.os.Looper;
import android.os.PowerManager;

import java.io.DataInputStream;
import java.io.IOException;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;

public class TcpServerService extends Service {
    public static final int SERVER_PORT = 6789;
    private boolean mStarted;
    private PowerManager.WakeLock mWakeLock;
    private ServerSocket mServerSocket;


    @Override
    public IBinder onBind(Intent intent) {
        return new TcpServerBinder();
    }

    @SuppressLint("InvalidWakeLockTag")
    @Override
    public void onCreate() {
        super.onCreate();
        PowerManager powerManager = (PowerManager) getSystemService(Context.POWER_SERVICE);
        mWakeLock = powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "TcpServerService");
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return START_STICKY;
    }

    private void startServer(final TcpServerListener listener) {
        mStarted = true;
        mWakeLock.acquire(30*60*1000L);
        final Looper myLooper = Looper.myLooper();

        new HandlerThread("TcpServerThread") {
            @Override
            protected void onLooperPrepared() {
                new Handler(getLooper()).post(new Runnable() {
                    @Override
                    public void run() {
                        mServerSocket = null;
                        try {
                            mServerSocket = new ServerSocket();
                            mServerSocket.setReuseAddress(true);
                            mServerSocket.setSoTimeout(0);
                            mServerSocket.bind(new InetSocketAddress(SERVER_PORT));

                            while (mStarted) {
                                try {
                                    Socket connectionSocket = mServerSocket.accept();
                                    onNewReceive(connectionSocket, myLooper, listener);
                                } catch (IOException e) {
                                    e.printStackTrace();
                                }
                            }
                        } catch (IOException e) {
                            e.printStackTrace();
                            listener.onStartFailure(e);
                        } finally {
                            if (mServerSocket != null && !mServerSocket.isClosed()) {
                                try {
                                    mServerSocket.close();
                                } catch (IOException e1) {
                                    e1.printStackTrace();
                                }
                            }
                            getLooper().quitSafely();
                        }
                    }
                });
            }
        }.start();
    }

    private void onNewReceive(final Socket connectionSocket,
                              final Looper myLooper,
                              final TcpServerListener listener) {
        new HandlerThread("ClientServingThread") {
            @Override
            protected void onLooperPrepared() {
                new Handler(getLooper()).post(new Runnable() {
                    @Override
                    public void run() {
                        byte[] bytes = null;
                        try {
                            DataInputStream dataInputStream = new DataInputStream(connectionSocket.getInputStream());

                            int length = dataInputStream.readInt();
                            if (length > 0) {
                                bytes = new byte[length];
                                dataInputStream.readFully(bytes, 0, bytes.length);
                            }
                            onReceive(myLooper, listener, bytes, connectionSocket.getInetAddress());
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                        getLooper().quitSafely();
                    }
                });
            }
        }.start();
    }

    private void onReceive(Looper looper,
                           final TcpServerListener listener,
                           final byte[] bytes,
                           final InetAddress inetAddress) {
        new Handler(looper).post(new Runnable() {
            @Override
            public void run() {
                listener.onReceive(bytes, inetAddress);
            }
        });
    }

    private void stopServer() {
        mStarted = false;
        mWakeLock.release();
        new HandlerThread("ServerTerminator") {
            @Override
            protected void onLooperPrepared() {
                new Handler(getLooper()).post(new Runnable() {
                    @Override
                    public void run() {
                        try {
                            mServerSocket.close();
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                    }
                });
            }
        }.start();
    }

    public class TcpServerBinder extends Binder {
        private TcpServerListener mListener;

        public void setListener(TcpServerListener listener) {
            mListener = listener;
        }

        public void startServer() {
            TcpServerService.this.startServer(mListener);
        }

        public void stopServer() {
            TcpServerService.this.stopServer();
        }
    }
}

File: near/src/main/java/com/adroitandroid/near/connect/server/TcpServerService.kt
package com.adroitandroid.near.connect.server

import android.annotation.SuppressLint
import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.*
import android.os.PowerManager.WakeLock
import java.io.DataInputStream
import java.io.IOException
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket

class TcpServerService : Service() {
    private var mStarted = false
    private lateinit var mWakeLock: WakeLock
    private var mServerSocket: ServerSocket? = null

    override fun onBind(intent: Intent): IBinder? {
        return TcpServerBinder()
    }

    @SuppressLint("InvalidWakeLockTag")
    override fun onCreate() {
        super.onCreate()
        val powerManager = getSystemService(Context.POWER_SERVICE) as PowerManager
        mWakeLock = powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "TcpServerService")
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        return START_STICKY
    }

    private fun startServer(port: Int, listener: TcpServerListener) {
        mStarted = true
        mWakeLock.acquire(30 * 60 * 1000L)
        val myLooper = Looper.myLooper() ?: Looper.getMainLooper()

        object : HandlerThread("TcpServerThread") {
            override fun onLooperPrepared() {
                Handler(looper).post {
                    mServerSocket = null
                    try {
                        mServerSocket = ServerSocket()
                        mServerSocket!!.reuseAddress = true
                        mServerSocket!!.soTimeout = 0
                        mServerSocket!!.bind(InetSocketAddress(port))
                        while (mStarted) {
                            try {
                                val connectionSocket = mServerSocket!!.accept()
                                onNewReceive(connectionSocket, myLooper, listener)
                            } catch (e: IOException) {
                                e.printStackTrace()
                            }
                        }
                    } catch (e: IOException) {
                        e.printStackTrace()
                        listener.onStartFailure(e)
                    } finally {
                        if (mServerSocket != null && !mServerSocket!!.isClosed) {
                            try {
                                mServerSocket!!.close()
                            } catch (e1: IOException) {
                                e1.printStackTrace()
                            }
                        }
                        looper.quitSafely()
                    }
                }
            }
        }.start()
    }

    private fun onNewReceive(connectionSocket: Socket,
                             myLooper: Looper,
                             listener: TcpServerListener) {

        object : HandlerThread("ClientServingThread") {
            override fun onLooperPrepared() {
                Handler(looper).post {
                    try {
                        val dataInputStream = DataInputStream(connectionSocket.getInputStream())
                        val length = dataInputStream.readInt()
                        if (length > 0) {
                            val bytes = ByteArray(length)
                            dataInputStream.readFully(bytes, 0, bytes.size)
                            onReceive(myLooper, listener, bytes, connectionSocket.inetAddress)
                        }
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }
                    looper.quitSafely()
                }
            }
        }.start()
    }

    private fun onReceive(looper: Looper,
                          listener: TcpServerListener,
                          bytes: ByteArray,
                          inetAddress: InetAddress) {
        Handler(looper).post { listener.onReceive(bytes, inetAddress) }
    }

    private fun stopServer() {
        mStarted = false
        mWakeLock.release()
        object : HandlerThread("ServerTerminator") {
            override fun onLooperPrepared() {
                Handler(looper).post {
                    try {
                        mServerSocket?.close()
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }
                }
            }
        }.start()
    }

    inner class TcpServerBinder : Binder() {
        private var mListener: TcpServerListener? = null
        private var mPort: Int = SERVER_PORT

        fun setListener(listener: TcpServerListener) {
            mListener = listener
        }

        fun setPort(port: Int) {
            mPort = port
        }

        fun startServer() {
            this@TcpServerService.startServer(mPort, mListener!!)
        }

        fun stopServer() {
            this@TcpServerService.stopServer()
        }
    }

    companion object {
        const val SERVER_PORT = 6789
    }
}